resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/DriverInitContainerComponentsProvider.scala

```diff
@@ -128,28 +129,36 @@ private[spark] class DriverInitContainerComponentsProviderImpl(
           maybeResourceStagingServerInternalTrustStorePassword,
           maybeResourceStagingServerInternalTrustStoreType,
           INIT_CONTAINER_SECRET_VOLUME_MOUNT_PATH)
-    }
-    new SparkInitContainerConfigMapBuilderImpl(
+      }
+      new SparkInitContainerConfigMapBuilderImpl(
         sparkJars,
         sparkFiles,
         jarsDownloadPath,
         filesDownloadPath,
         configMapName,
         configMapKey,
         submittedDependencyConfigPlugin)
+    } else new SparkInitContainerEmptyConfigMapBuilderImpl()
   }
 
   override def provideContainerLocalizedFilesResolver(): ContainerLocalizedFilesResolver = {
     new ContainerLocalizedFilesResolverImpl(
         sparkJars, sparkFiles, jarsDownloadPath, filesDownloadPath)
   }
+  val containerLocalizedFilesResolver = provideContainerLocalizedFilesResolver()
+  val resolvedSparkJars = containerLocalizedFilesResolver.resolveSubmittedSparkJars()
+  val resolvedSparkFiles = containerLocalizedFilesResolver.resolveSubmittedSparkFiles()
+  // Bypass init-containers if `spark.jars` and `spark.files` is empty or only has `local://` URIs
+  private val needInitContainer = (resolvedSparkJars ++ resolvedSparkFiles).exists { e =>
+    e.nonEmpty && !e.startsWith("local://")
+  }
 
   override def provideExecutorInitContainerConfiguration(): ExecutorInitContainerConfiguration = {
-    new ExecutorInitContainerConfigurationImpl(
+    if (needInitContainer) new ExecutorInitContainerConfigurationImpl(
```

Comment:
Difficulty of this statement is that for PySpark and SparkR there will be files that arent necessarily SparkJars or SparkFiles... so this logic wouldn't work specifically given PR [351](https://github.com/apache-spark-on-k8s/spark/pull/351) As @mccheah has noted, this may require re-structuring. Good to bring up at next SIG meeting